### PR TITLE
gccWithoutTargetLibc: if glibc, enable threads and pass headers

### LINKED ISSUE #213453: `mbuffer`: cross-compiled libgcc_s.so.1 must be built with glibc headers for pthread_cleanup_push() to work
### Describe the bug

On `armv7l-linux` or `aarch64-linux` cross-compiled from `x86_64-linux`, `mbuffer` breaks:

```
$ echo "hello" | mbuffer
hello
libgcc_s.so.1 must be installed for pthread_exit to work
Aborted (core dumped)
```

From `armv7l-linux`:

```
Module linux-vdso.so.1 with build-id 04745a31932e964ec4cd2f2af22bff73c959bc5d
Module libatomic.so.1 without build-id.
Module libdl.so.2 with build-id e21ee3b08115e4ff693e068a5ae47e30a6eed56e
Module ld-linux-armhf.so.3 with build-id 5450bffccfed072763370bd64a69b6ea758babb1
Module libc.so.6 with build-id cc4343a8abc2fde08e30176f06b9f4385f438ef5
Module libpthread.so.0 with build-id 6b805729d47513ad23db66eeac106a69998cb666
Module libm.so.6 with build-id 9350661951669899d3d720161626db749d157003
Module libcrypto.so.3 with build-id f4a6e190dcfd9af5793d9afd5cd89ef82559a3e4
Module mbuffer without build-id.
Stack trace of thread 2019:
#0  0x00000000b6a7b2f8 __pthread_kill_implementation (libc.so.6 + 0x7b2f8)
ELF object binary architecture: ARM
```

### Steps To Reproduce

You can either run this on another architecture using `boot.binfmt.emulatedSystems = [ "armv7l-linux" "aarch64-linux" ]`.

```console
$ nix build nixpkgs#pkgsCross.aarch64-multiplatform.mbuffer
$ echo "hello" | result/bin/mbuffer
```

or

```console
$ nix build nixpkgs#pkgsCross.armv7l-hf-multiplatform.mbuffer
$ echo "hello" | result/bin/mbuffer
```

### Expected behavior

`mbuffer` should not crash:

```
$ echo "hello" | mbuffer
hello
summary:  0.0 kiByte in  0.0sec - average of  0.0 kiB/s
```

### Additional context

This was responsible for `syncoid` failing to replicate datasets to my ARM NAS.

The issue does _not_ occur for natively compiled `aarch64-linux` builds. It appears to be a cross-compilation issue only.

### Notify maintainers

@tokudan @skeuchel 

### Metadata

```console
 - system: `"armv7l-linux"`
 - host os: `Linux 6.1.8, NixOS, 22.11 (Raccoon), 22.11.20230127.cc4bb87`
```

### PR DESCRIPTION
**Pull request by ghost**
:warning: see update [below](https://github.com/NixOS/nixpkgs/pull/241208#issuecomment-1624164228) :warning:

###### Description of changes

If `${targetPackages.glibc}/include/stdio.h` does not exist during the `configurePhase` of `pkgsCross.*.gcc.libgcc`, the stack unwinder will be silently disabled.

The resulting libgcc_s.so will deliberately crash any process which attempts to use `pthread_cleanup_push()` or similar calls.

This PR ensures that `${targetPackages.glibc}/include/stdio.h` *does* exist during the `configurePhase` of `pkgsCross.*.gcc.libgcc`, so we get an unwinder-capable `libgcc_s.so`.

This also adds a new package `glibcHeaders`, which allows to build the glibc headers without first having built a targetPlatform compiler.  This is needed in order to break the dependency cycle between glibc and libgcc.  Additional details are in the commit message for that particular commit

##### Closes:

- https://github.com/NixOS/nixpkgs/issues/213453

##### Includes

- https://github.com/NixOS/nixpkgs/pull/241206
- https://github.com/NixOS/nixpkgs/pull/238154

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by amjoseph-nixpkgs** [comment]
Update: so, this does fix #213453, and it does work on `x86_64` and `aarch64`.  Unfortunately it pretty much doesn't work on any other architecture.

Apparently being able to `./configure` without a targetPlatform compiler is a fairly new feature for `glibc` and it has only been implemented on the popular platforms.  So I don't think this PR will ever be undraftified (and, because of that, I am unlikely to hunt down the OfBorg eval failure although I can rebuild my whole userspace without hitting it).

After much digging around in the guts of `glibc` I think I've found a better solution: right now we use the `libgcc` from the *first-stage* `gcc`.  That `libgcc` is missing the unwinder.  All we need to do is use the `libgcc` from the *second-stage* gcc instead.  There are two ways to do this:

1. The easy way: add an extra recompile of glibc for `pkgsCross.*.stdenv`.  This is simple enough that it might be backport-eligible.

2. The better way: use a technique similar to `patchelf --add-rpath` to update the pointer in `glibc`.  This gives the same result without an extra recompile.  Most importantly, this is a change which could be applied to both `pkgsCross.*.stdenv` *and also* to native-stdenv.  This will take time, testing, and have to go through staging.  We can't use `patchelf` since the "pointer" isn't an ELF `DT_NEEDED`, but it is a simple string in the static-data section of `ld.so`.

So to summarize:

1. This PR is 99%-likely to be abandoned.
2. I will write a PR for strategy #1 above in the next week or two, and if merged propose it for backporting.
3. I will write another PR which reverts strategy #1 and implements strategy #2 sometime in the next few months.  This PR will not be proposed for backporting to 23.05.
